### PR TITLE
[Messenger] add support for abstract handlers

### PR DESCRIPTION
**Pull request by timiTao**
| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | no
| New feature?  | yes 
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | #31417    
| License       | MIT
| Doc PR        | symfony/symfony-docs#... 

Added handling abstract handler.

### CODE REVIEW

File: src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php [hunk 1]
```diff
@@ -176,12 +183,10 @@ private function registerHandlers(ContainerBuilder $container, array $busIds)
 
         foreach ($busIds as $bus) {
             $container->register($locatorId = $bus.'.messenger.handlers_locator', HandlersLocator::class)
-                ->setArgument(0, $handlersLocatorMappingByBus[$bus] ?? [])
-            ;
+                ->setArgument(0, $handlersLocatorMappingByBus[$bus] ?? []);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php]: please revert these, they are unrelated

File: src/Symfony/Component/Messenger/Tests/DependencyInjection/MessengerPassTest.php [hunk 2]
```diff
@@ -208,6 +209,61 @@ public function testGetClassesAndMethodsAndPrioritiesFromTheSubscriber()
...
+        $this->assertHandlerDescriptor(
+            $container,
+            $messageHandlerMapping,
+            DummyMessage::class,
+            [$abstractHandlerId]
+        );
+
+//        /////
+//        ///
+//        $container = $this->getContainerBuilder($eventsBusId = 'event_bus');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Tests/DependencyInjection/MessengerPassTest.php]: should be removed or uncommented

File: src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php [hunk 3]
```diff
@@ -79,7 +79,14 @@ private function registerHandlers(ContainerBuilder $container, array $busIds)
                     throw new RuntimeException(sprintf('Invalid handler service "%s": bus "%s" specified on the tag "%s" does not exist (known ones are: %s).', $serviceId, $tag['bus'], $this->handlerTag, implode(', ', $busIds)));
                 }
 
-                $className = $container->getDefinition($serviceId)->getClass();
+                $serviceIdDefinition = $container->getDefinition($serviceId);
+
+                if ($serviceIdDefinition instanceof ChildDefinition) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php]: this should be in a "while" loop as the parent can itself be a ChildDefinition

**Review by nicolas-grekas — COMMENTED** [review]
I'd be OK to add this for 4.3 - here are some comments before.

File: src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php (hunk 1 above)
Comment (timiTao) [reply src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php]: This was made by the bot, after the apply patch.

**Comment by nicolas-grekas** [comment]
Can you please `rebase -i` to get rid of the extra commits?

File: src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php [hunk 4]
```diff
@@ -287,8 +294,8 @@ private function registerReceivers(ContainerBuilder $container, array $busIds)
     private function registerBusToCollector(ContainerBuilder $container, string $busId)
     {
         $container->setDefinition(
-            $tracedBusId = 'debug.traced.'.$busId,
-            (new Definition(TraceableMessageBus::class, [new Reference($tracedBusId.'.inner')]))->setDecoratedService($busId)
+            $tracedBusId = 'debug.traced.' . $busId,
+            (new Definition(TraceableMessageBus::class, [new Reference($tracedBusId . '.inner')]))->setDecoratedService($busId)
```
Comment (sroze) [inline src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php]: Can you revert these style changes?

File: src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php [hunk 5]
```diff
@@ -79,7 +79,14 @@ private function registerHandlers(ContainerBuilder $container, array $busIds)
                     throw new RuntimeException(sprintf('Invalid handler service "%s": bus "%s" specified on the tag "%s" does not exist (known ones are: %s).', $serviceId, $tag['bus'], $this->handlerTag, implode(', ', $busIds)));
                 }
 
-                $className = $container->getDefinition($serviceId)->getClass();
+                $serviceIdDefinition = $container->getDefinition($serviceId);
```
Comment (sroze) [inline src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php]: Can you put this logic (including the `while` loop) into a private method? I'd name it `getDefinitionClass(string $serviceId): string;`

**Review by sroze — CHANGES_REQUESTED** [review]

**Comment by sroze** [comment]
@timiTao can you rebase to get rid of these extra commits in your branch please?

File: src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php [hunk 6]
```diff
@@ -321,4 +321,17 @@ private function registerBusMiddleware(ContainerBuilder $container, string $busI
 
         $container->getDefinition($busId)->replaceArgument(0, new IteratorArgument($middlewareReferences));
     }
+
+    private function getDefinitionClass(string $serviceId, ContainerBuilder $container)
```
Comment (juanmiguelbesada) [inline src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php]: I think is better set $container as first parameter (to keep consistent with other methods)

```suggestion
    private function getDefinitionClass(ContainerBuilder $container, string $serviceId)
```

Comment (juanmiguelbesada) [inline src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php]:
```suggestion
    private function getDefinitionClass(ContainerBuilder $container, string $serviceId)
```

File: src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php [hunk 7]
```diff
@@ -79,7 +79,7 @@ private function registerHandlers(ContainerBuilder $container, array $busIds)
                     throw new RuntimeException(sprintf('Invalid handler service "%s": bus "%s" specified on the tag "%s" does not exist (known ones are: %s).', $serviceId, $tag['bus'], $this->handlerTag, implode(', ', $busIds)));
                 }
 
-                $className = $container->getDefinition($serviceId)->getClass();
+                $className = $this->getDefinitionClass($serviceId, $container);
```
Comment (juanmiguelbesada) [inline src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php]:
```suggestion
                $className = $this->getDefinitionClass($container, $serviceId);
```

**Review by juanmiguelbesada — COMMENTED** [review]
In my opinion, I think is better set $container as first parameter of getDefinitionClass method (to keep consistent with other methods)

File: src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php [hunk 8]
```diff
@@ -321,4 +321,17 @@ private function registerBusMiddleware(ContainerBuilder $container, string $busI
...
     }
+
+    private function getDefinitionClass(ContainerBuilder $container, string $serviceId)
+    {
+        $serviceIdDefinition = $container->getDefinition($serviceId);
+        while ($serviceIdDefinition instanceof ChildDefinition) {
+            if (null !== $serviceIdDefinition->getClass()) {
+                break;
+            }
+            $serviceIdDefinition = $container->getDefinition($serviceIdDefinition->getParent());
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php]: we should use findDefinition here instead of getDefinition to be extra safe

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @timiTao.

**Pull request merged**